### PR TITLE
Flink: Fix range distribution npe when value is null

### PR DESCRIPTION
**Pull request by Guosmilesmile**
When configuring the distribution mode to RANGE, if the partition field in the data contains null values, it will cause the SortKey serialization to fail, resulting in the job continuously restarting.

If we set the partition as `bucket(5, name)`, and some data has a null value for `name`, then the `DataStatisticsOperator` will throw an error when serializing the statistics and sending them to the coordinator.

In SortKeySerializer.serialize
```
  public void serialize(SortKey record, DataOutputView target) throws IOException {

    for (int i = 0; i < size; ++i) {
      int fieldId = transformedFields[i].fieldId();
      Type.TypeID typeId = transformedFields[i].type().typeId();
      switch (typeId) {
        case BOOLEAN:
          target.writeBoolean(record.get(i, Boolean.class));
          break;
        case INTEGER:
        case DATE:
          target.writeInt(record.get(i, Integer.class));
          break;
        
          ...
       }


```

Error log:

```
Caused by: org.apache.flink.util.SerializedThrowable: java.lang.NullPointerException
    at org.apache.iceberg.flink.sink.shuffle.SortKeySerializer.serialize(SortKeySerializer.java:133) ~[flink-onejar-4.0.0.jar:?]
    at org.apache.iceberg.flink.sink.shuffle.SortKeySerializer.serialize(SortKeySerializer.java:49) ~[flink-onejar-4.0.0.jar:?]
    at org.apache.flink.api.common.typeutils.base.MapSerializer.serialize(MapSerializer.java:136) ~[flink-dist-1.19.1.jar:1.19.1]
    at org.apache.iceberg.flink.sink.shuffle.DataStatisticsSerializer.serialize(DataStatisticsSerializer.java:113) ~[flink-onejar-4.0.0.jar:?]
    at org.apache.iceberg.flink.sink.shuffle.DataStatisticsSerializer.serialize(DataStatisticsSerializer.java:38) ~[flink-onejar-4.0.0.jar:?]
    at org.apache.iceberg.flink.sink.shuffle.StatisticsUtil.serializeDataStatistics(StatisticsUtil.java:46) ~[flink-onejar-4.0.0.jar:?]
    at org.apache.iceberg.flink.sink.shuffle.StatisticsEvent.createTaskStatisticsEvent(StatisticsEvent.java:51) ~[flink-onejar-4.0.0.jar:?]
    at org.apache.iceberg.flink.sink.shuffle.DataStatisticsOperator.snapshotState(DataStatisticsOperator.java:227) ~[flink-onejar-4.0.0.jar:?]
    at org.apache.flink.streaming.api.operators.StreamOperatorStateHandler.snapshotState(StreamOperatorStateHandler.java:234) ~[flink-dist-1.19.1.jar:1.19.1]
```

We think If a field is set as a partition and a transformation function is also applied, the scenario will inevitably occur when the field contains null values. I believe that during the statistics, we can ignore the data with null values in the relevant partition field. If the sample size is small, it should not affect the job's execution. If the sample size is large, using the default round strategy would also be appropriate.

This PR mainly traverses the contents of the partition field and skips the statistics collection for any parts that contain null values.

### CODE REVIEW

**Comment by ConeyLiu** [comment]
Thanks for the contributions. Should we better handle the null value instead of skipping it? Could you also add the UT to cover these changes?
cc @stevenzwu @pvary

**Comment by Guosmilesmile** [comment]
@ConeyLiu yes，I have added UT to cover the relevant changes.

**Comment by pvary** [comment]
Seems like a hacky solution to me. We will miss all of the advantages for the range distribution mode for records which are affected. Could we just serialize the `null` values correctly instead?

**Comment by Guosmilesmile** [comment]
@pvary  Yes, you are right. We changed our approach to handle null values instead of filtering them out. Before serialization, we added a flag for each field, where true indicates that the field is null. During serialization and deserialization, we read the flag first, which allows us to correctly identify null. The related code has been updated; please take a look to see if it is feasible.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java [hunk 1]
```diff
@@ -124,6 +124,13 @@ public void serialize(SortKey record, DataOutputView target) throws IOException
     for (int i = 0; i < size; ++i) {
       int fieldId = transformedFields[i].fieldId();
       Type.TypeID typeId = transformedFields[i].type().typeId();
+      Object value = record.get(i, Object.class);
+      if (value == null) {
+        target.writeBoolean(true);
+        continue;
+      } else {
+        target.writeBoolean(false);
+      }
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:133]: nit: newline after closing a block and starting another statement

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java [hunk 2]
```diff
@@ -192,6 +199,11 @@ public SortKey deserialize(SortKey reuse, DataInputView source) throws IOExcepti
         reuse.size(),
         size);
     for (int i = 0; i < size; ++i) {
+      boolean isNull = source.readBoolean();
+      if (isNull) {
+        reuse.set(i, null);
+        continue;
+      }
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:206]: nit: newline after closing a block and starting another statement

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestDataStatisticsOperator.java [hunk 3]
```diff
@@ -136,6 +137,27 @@ public void testProcessElement(StatisticsType type) throws Exception {
...
+          operator.processElement(new StreamRecord<>(GenericRowData.of(null, 5)));
+          operator.processElement(new StreamRecord<>(GenericRowData.of(StringData.fromString("a"), 3)));
+
+          DataStatistics localStatistics = operator.localStatistics();
+          SortKeySerializer sortKeySerializer = new SortKeySerializer(Fixtures.SCHEMA, Fixtures.SORT_ORDER);
+          DataStatisticsSerializer taskStatisticsSerializer = new DataStatisticsSerializer(sortKeySerializer);
+          DataOutputSerializer outputView = new DataOutputSerializer(1024);
+
+          taskStatisticsSerializer.serialize(localStatistics, outputView);
+          testHarness.endInput();
```
Comment (pvary) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestDataStatisticsOperator.java:157]: We should check if the resulting statistics is correct

**Comment by pvary** [comment]
@Guosmilesmile: Thanks for the changes. Left some comments, but started the tests to see if this change cause any other test failures.

Please remove the 1.19, 1.18 changes for now - we usually cherry pick the changes to the other versions after the original PR for the main versions has been merged. This is better for the reviewer (smaller number of files), and better for the contributor (if there is a change request during the review, they don't have to keep the different versions in sync)

It would be nice to have an end2end test for null values too.
Currently we only tests that the statistics are collected correctly, but there might be some issues when applying the stats. It would be nice to have a test for this case too.

**Comment by pvary** [comment]
@Guosmilesmile: Please apply spotless to remove the formatting failures:
```
./gradlew spotlessApply
```

**Comment by Guosmilesmile** [comment]
@pvary  Thank you very much for your guidance. I have made the changes according to your suggestions. Please take a look and see if they are feasible.

**Comment by pvary** [comment]
Thanks for the changes @Guosmilesmile!
One question remains from my side:

> It would be nice to have an end2end test for null values too.
> Currently we only tests that the statistics are collected correctly, but there might be some issues when applying the stats. It would be nice to have a test for this case too.

**Comment by Guosmilesmile** [comment]
@pvary I have added test cases for DataStatisticsOperator and DataStatisticsCoordinator, specifically for the null scenario. I verified the ProcessElement and EventHand scenarios, simulating the process of data being processed by the operation and receiving statistical information that contains null values for handling. Could you please take a look and see if this is feasible? Thank you!

**Comment by Guosmilesmile** [comment]
@pvary I have also added a unit test for the scenario of building from the table to the sink, which includes data containing null values. Could you please review it as well? Thank you very much!

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestDataStatisticsCoordinator.java [hunk 4]
```diff
@@ -152,6 +152,80 @@ public void testDataStatisticsEventHandling(StatisticsType type) throws Exceptio
...
   }
 
+  @ParameterizedTest
+  @EnumSource(StatisticsType.class)
+  public void testDataStatisticsEventHandlingWithNullValue(StatisticsType type) throws Exception {
+    try (DataStatisticsCoordinator dataStatisticsCoordinator = createCoordinator(type)) {
+      dataStatisticsCoordinator.start();
+      tasksReady(dataStatisticsCoordinator);
+
+      SortKey NullSortKey = Fixtures.SORT_KEY.copy();
```
[the remark this answers is not in the document]
Comment (Guosmilesmile) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestDataStatisticsCoordinator.java]: @pvary I have change it , please check it, thank you

[the remark this answers is not in the document]
Comment (Guosmilesmile) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/sink/shuffle/TestDataStatisticsCoordinator.java]: @pvary I'm sorry, there was a little mistake just now. Please try again. Thank you very much.

**Comment by Guosmilesmile** [comment]
@pvary I'm sorry, I just re-uploaded the missing parts. Please help me trigger the test again. I apologize for the inconvenience.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java [hunk 5]
```diff
@@ -124,6 +124,14 @@ public void serialize(SortKey record, DataOutputView target) throws IOException
     for (int i = 0; i < size; ++i) {
       int fieldId = transformedFields[i].fieldId();
       Type.TypeID typeId = transformedFields[i].type().typeId();
+      Object value = record.get(i, Object.class);
```
Comment (stevenzwu) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:127]: this will add a boolean flag overhead for every field. but I can't think of any better alternative. so this is fine to me.

However, ideally we need to figure out how to evolve it in a compatible way. probably leverage `TypeSerializerSnapshot`

Comment (Guosmilesmile) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:127]: Can we accept this method to fix the bug for now?

Comment (Guosmilesmile) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:127]: @stevenzwu  I have looked at the current implementation in Flink, and the serialization for null values references the implementation of NullableSerializer. The underlying approach is also to add a boolean flag at the beginning of the field to distinguish, which is quite similar to the approach in this PR. Can we use the current method to fix this issue for now, or are there any other better handling methods?


```java
NullableSerializer<T> extends TypeSerializer<T>

    @Override
    public void serialize(T record, DataOutputView target) throws IOException {
        if (record == null) {
            target.writeBoolean(true);
            target.write(padding);
        } else {
            target.writeBoolean(false);
            originalSerializer.serialize(record, target);
        }
    }

    @Override
    public T deserialize(DataInputView source) throws IOException {
        boolean isNull = deserializeNull(source);
        return isNull ? null : originalSerializer.deserialize(source);
    }
```

Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:127]: @stevenzwu: What do you think about a serializer version, something like the SimpleVersionedSerializer uses. Maybe don't store the version for every key, but inherit the version from the statistics object?

**Comment by Guosmilesmile** [comment]
@pvary Because a new flag has been added, the testSerializationSize test case needs to be modified to reflect the corresponding size value + 1byte for the flag . It has been adjusted. I have run the tests locally myself, and all unit tests have passedand I would appreciate it if you could start the tests again.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java (hunk 5 above)
Comment (mxm) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:127]: Nulls are not allowed everywhere in Flink. Primitive types generally don't support null. In other places, Flink uses boolean flags to encode nulls, e.g. in PojoSerializer: https://github.com/apache/flink/blob/b9c92371dba07b6bfe4368d7b6d7f7c575b4c603/flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java#L392

KeyBy operations are not allowed on null values.

+1 to Steven's suggestion to version the serializer. We already have [SortKeySerializerSnapshot](https://github.com/apache/iceberg/blob/163e2068f96f139632488f36928bf443c9be326f/flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java#L278). We just need to add a version check for the new boolean serializer because this is an incompatible change which requires serializer migration.

**Review by mxm — COMMENTED** [review]
The approach here generally looks good to me once we address the remaining comments.

**Comment by Guosmilesmile** [comment]
@mxm Thank you very much for your suggestions. I need to add a version check in SortKeySerializerSnapshot. If the state is restored from an old version, I will directly return TypeSerializerSchemaCompatibility.incompatible(). Is this approach feasible?

**Comment by Guosmilesmile** [comment]
@mxm @stevenzwu @pvary Thank you all for your suggestions. I have submitted a version that mainly modifies the SortKeySerializerSnapshot and implements version detection. If the version does not match, it will return the corresponding message to avoid parsing exceptions and data errors when restoring from an old state to a new state. I would appreciate it if you could take a look and see if this modification is feasible

**Comment by mxm** [comment]
@Guosmilesmile Ideally, we want to return compatibleAfterMigration() and make sure the old and new serializer can be instantiated.

**Comment by Guosmilesmile** [comment]
@mxm I'm sorry, I'm a bit confused and would like to ask for your advice. Which part do I need to modify?

Based on the current modifications, simply change TypeSerializerSchemaCompatibility.incompatible() to compatibleAfterMigration().
Create a new NullAbleSortKeySerializer, keep the SortKeySerializer, replace all calls to SortKeySerializer, and in resolveSchemaCompatibility, return compatibleAfterMigration() after checking the type of oldSerializerSnapshot.
Which option should I choose? Please provide detailed advice. Thank you very much.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java [hunk 6]
```diff
@@ -310,10 +324,16 @@ public void writeSnapshot(DataOutputView out) throws IOException {
...
-      } else {
-        throw new IllegalArgumentException("Unknown read version: " + readVersion);
+      switch (readVersion) {
+        case 1:
+          throw new UnsupportedOperationException(
+              String.format(
+                  "No longer supported version [%s]. Please upgrade first . ", readVersion));
+        case 2:
+          readV2(in);
+          break;
```
Comment (mxm) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:334]: Reading the snapshot isn't any different for V1/V2. The `readV1()` or `readV2()` methods are identical. We can just keep the current one. However, we need to save the serializer version as an `int` field which we can read in `restoreSerializer()`, just like the schema and the sort order. We can then serialize/deserialize depending on the version used.

For example, in the serializer we would do:
```
if (version == 1) {
  // Don't read/write null indicator
} else {
  // Read/write null indicator
}

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java [hunk 7]
```diff
@@ -324,6 +344,10 @@ public TypeSerializerSchemaCompatibility<SortKey> resolveSchemaCompatibility(
         return TypeSerializerSchemaCompatibility.incompatible();
       }
 
+      if (oldSerializerSnapshot.getCurrentVersion() != this.getCurrentVersion()) {
+        return TypeSerializerSchemaCompatibility.incompatible();
+      }
```
Comment (mxm) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:349]:
```suggestion
      if (oldSerializerSnapshot.getCurrentVersion() == 1 && this.getCurrentVersion() == 2) {
        return TypeSerializerSchemaCompatibility.compatibleAfterMigration();
      }
```

**Review by mxm — COMMENTED** [review]
Hey @Guosmilesmile! Please see the suggestions inline. Let me know if you have questions.

**Comment by Guosmilesmile** [comment]
Hi @mxm! Thank you very much for your suggestions. Following your advice, I added a version number to SortKeySerializer through restoreSerializer, implementing different serialization methods for v1 and v2. Based on the performance in the test job, the TaskManager chieve state compatibility. However, the JobManager fails to recover due to the DataStatisticsCoordinator, which leads to a failure to start because the SortKeySerializer used in CompletedStatisticsSerializer and GlobalStatisticsSerializer is instantiated directly. This results in a failure when trying to parse the v1 state with the v2 version. Additionally, these two serializers do not change after the Coordinator is initialized. Given this scenario, do you have any good suggestions?

**Comment by mxm** [comment]
Thanks for the update @Guosmilesmile! Unfortunately, we don't have a way to encode the serializer version for all serializers, so a best-effort approach to retry with a different serializer version on restore failure may be the best we can do.

**Comment by Guosmilesmile** [comment]
@mxm Thank you very much for your suggestions. I have made the necessary modifications, and I appreciate you taking the time out of your busy schedule to review it again. I am very grateful for your guidance.

**Comment by Guosmilesmile** [comment]
@mxm Thank you very much for your suggestions. I have made the necessary modifications, and I appreciate you taking the time out of your busy schedule to review it again. I am very grateful for your guidance.

**Comment by Guosmilesmile** [comment]
@mxm @pvary @stevenzwu Thank you all very much for your guidance. I would appreciate it if you could take another look and let me know what needs to be done next to keep things moving forward.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/CompletedStatistics.java [hunk 8]
```diff
@@ -108,4 +108,20 @@ boolean isEmpty() {
...
+      }
+    } else {
+      if (null == keyFrequency()) {
+        return false;
+      }
+      if (keyFrequency().values().contains(null)) {
+        return false;
+      }
+    }
+    return true;
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/CompletedStatistics.java:125]: nit: newline before this line

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsUtil.java [hunk 9]
```diff
@@ -73,12 +73,24 @@ static byte[] serializeCompletedStatistics(
...
     try {
       DataInputDeserializer input = new DataInputDeserializer(bytes);
-      return statisticsSerializer.deserialize(input);
-    } catch (IOException e) {
-      throw new UncheckedIOException("Fail to deserialize aggregated statistics", e);
+      CompletedStatistics completedStatistics = statisticsSerializer.deserialize(input);
+      if (!completedStatistics.isValid()) {
+        throw new RuntimeException("Fail to deserialize aggregated statistics,change to v1");
+      }
+      return completedStatistics;
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsUtil.java:83]: nit: newline before this line

Comment (Guosmilesmile) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsUtil.java:84]: @pvary Thank you very much for your suggestions. I have made the corresponding changes based on your advice

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsUtil.java [hunk 10]
```diff
@@ -73,12 +73,25 @@ static byte[] serializeCompletedStatistics(
...
+    } catch (Exception e) {
+      try {
+        statisticsSerializer.changeSortKeySerializerVersion(1);
+        DataInputDeserializer input = new DataInputDeserializer(bytes);
+        CompletedStatistics deserialize = statisticsSerializer.deserialize(input);
+        statisticsSerializer.changeSortKeySerializerVersionLatest();
+        return deserialize;
+      } catch (IOException ioException) {
+        throw new UncheckedIOException("Fail to deserialize aggregated statistics", ioException);
+      }
```
Comment (mxm) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/StatisticsUtil.java:94]: We could add a comment here, explaining why switching versions was necessary.

**Review by mxm — APPROVED** [review]
I think this looks good, but I'll defer to @stevenzwu and @pvary for the final approval.

**Comment by Guosmilesmile** [comment]
> I think this looks good, but I'll defer to @stevenzwu and @pvary for the final approval.

@mxm Thank you very much for taking the time to review my code. The relevant comments have been added. 

@pvary @stevenzwu I would appreciate it if both of you could take a look for the final approval.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java [hunk 11]
```diff
@@ -52,8 +52,32 @@ class SortKeySerializer extends TypeSerializer<SortKey> {
...
+      Type resultType = sortField.transform().getResultType(sourceField.type());
+      Types.NestedField transformedField =
+          Types.NestedField.of(
+              sourceField.fieldId(),
+              sourceField.isOptional(),
+              sourceField.name(),
+              resultType,
+              sourceField.doc());
+      transformedFields[i] = transformedField;
+    }
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:78]: This code part seems to be a duplicate for me?
What is the reason for not using the same underlying method for the 2 constructors?

Comment (Guosmilesmile) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:78]: @pvary Thank you very much for your suggestion. Since all member variables are final, they can only be initialized through the constructor. Most of the initialization for SortKeySerializer uses the latest version, and only when restoring the serializer do we need to pass in the version number during initialization. Therefore, I added a new constructor. 

Another option is to remove the newly added constructor and use a setVersion method for injection. Because the version number is only needed for initialization in one place.

Which approach do you think is more appropriate?

Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java:78]: Could this work?
```
  SortKeySerializer(Schema schema, SortOrder sortOrder) {
    this(schema, sortOrder, SortKeySerializerSnapshot.CURRENT_VERSION);
  }
```

Comment (Guosmilesmile) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/SortKeySerializer.java]: @pvary Yes, it's feasible. I've made the adjustments. Could you please take another look to see if there are any other issues? Thanks ！

**Comment by pvary** [comment]
@stevenzwu: any last minute comments?

**Review by pvary — APPROVED** [review]

**Comment by pvary** [comment]
Thanks @Guosmilesmile for the PR, and @mxm and @stevenzwu for the review!

@Guosmilesmile please prepare the backport commits to the other Flink versions.

Thanks, Peter

**Comment by Guosmilesmile** [comment]
> Thanks @Guosmilesmile for the PR, and @mxm and @stevenzwu for the review!
> 
> @Guosmilesmile please prepare the backport commits to the other Flink versions.
> 
> Thanks, Peter

Thank you all for your reviews @mxm @stevenzwu @pvary . 

@pvary The related backport has been submitted, see #11745. Thank you!
